File: Laboratorio09/Propuestos/Ejercicio02/PythonExample/cuentas.py
import mysql.connector
from mysql.connector import Error

def main():
    try:
        connection = mysql.connector.connect(
            host='localhost',
            database='cuentasbancarias',
            user='usuario',
            password='password'
        )

        if connection.is_connected():
            print("Conexion exitosa a la base de datos")

            cursor = connection.cursor()

            # Desactivar autocommit para iniciar la transaccion
            connection.autocommit = False

            # 1. Insertar banco
            cursor.execute("INSERT INTO bancos (nombre, direccion) VALUES (%s, %s)", 
                           ("Banco Python", "Av. Central 789"))
            banco_id = cursor.lastrowid
            print("1er INSERT en tabla 'bancos'")
            print("ID del banco insertado:", banco_id)

            # 2. Insertar titular
            cursor.execute("INSERT INTO titulares (dni, nombre) VALUES (%s, %s)", 
                           ("99998888", "Grupo3"))
            print("2do INSERT en tabla 'titulares'")

            # 3. Insertar cuenta
            cursor.execute("INSERT INTO cuentas (dni_titular, banco_id, saldo) VALUES (%s, %s, %s)", 
                           ("99998888", banco_id, 2500.0))
            print("3er INSERT en tabla 'cuentas'")

            # 4. Intentar actualizar una cuenta inexistente (simula error)
            cursor.execute("UPDATE cuentas SET saldo = saldo - %s WHERE id = %s", 
                           (100.0, 999))  # ID 999 no existe

            if cursor.rowcount == 0:
                raise Exception("La cuenta a debitar no existe. Simulando error para rollback.")

            # Confirmar la transaccion si todo salio bien
            connection.commit()
            print("Transaccion completada con exito.")

    except Exception as e:
        print(f"ERROR: {e}")
        if connection:
            connection.rollback()
            print("Rollback ejecutado.")

    finally:
        if connection.is_connected():
            cursor.close()
            connection.close()
            print("Conexion cerrada.")

if __name__ == "__main__":
    main()

File: Laboratorio09/Propuestos/Ejercicio02/C++Example/cuentas.cpp
#include <iostream>
#include <memory>
#include <mysqlx/xdevapi.h>

using namespace std;
using namespace mysqlx;

int main() {
    Session sess;
    try {
        // Conexion a la base de datos
        Session sess("localhost", 33060, "usuario", "password", "cuentasbancarias");
        sess.startTransaction();  // iniciar transaccion

        // Insertar banco y recuperar ID
        Schema db = sess.getSchema("cuentasbancarias");
        Table bancos = db.getTable("bancos");
        Result res1 = bancos.insert("nombre", "direccion")
                             .values("Banco 3", "Av. Central 789")
                             .execute();
        int bancoId = static_cast<int>(res1.getAutoIncrementValue());
        cout << "Banco insertado. ID generado: " << bancoId << endl;

        // Insertar titular
        Table titulares = db.getTable("titulares");
        titulares.insert("dni", "nombre")
                  .values("99998888", "Grupo4")
                  .execute();
        cout << "Titular insertado." << endl;

        // Insertar cuenta
        Table cuentas = db.getTable("cuentas");
        cuentas.insert("dni_titular", "banco_id", "saldo")
               .values("99998888", bancoId, 2500.0)
               .execute();
        cout << "Cuenta insertada." << endl;

        // Simular error con actualizacion invalida
        Result res4 = cuentas.update()
                             .set("saldo", "saldo - 100")
                             .where("id = 999")
                             .execute();
        if (res4.getAffectedItemsCount() == 0) {
            throw runtime_error("Cuenta inexistente. Forzando rollback.");
        }

        sess.commit();  // confirmar transaccion
        cout << "Transaccion completada." << endl;

    } catch (const Error &err) {
        cerr << "ERROR: " << err.what() << endl;
        try {
            mysqlx::sess.rollback();
            cerr << "Rollback ejecutado." << endl;
        } catch (...) {
            cerr << "Error al hacer rollback." << endl;
        }
    }
    return 0;
}
